controllers/compaction/reconciler.go

```diff
@@ -154,6 +154,19 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	diff := delta - full
 	metricNumDeltaEvents.With(prometheus.Labels{druidmetrics.EtcdNamespace: etcd.Namespace}).Set(float64(diff))
 
+	// Update metrics for currently running compaction job, if any
+	job := &batchv1.Job{}
+	err = r.Get(ctx, types.NamespacedName{Name: etcd.GetCompactionJobName(), Namespace: etcd.Namespace}, job)
+	if err != nil {
+		logger.Info("Could not fetch any running compaction job")
+	}
+
+	if job != nil && job.Name != "" {
+		metricJobsCurrent.With(prometheus.Labels{druidmetrics.EtcdNamespace: etcd.Namespace}).Set(1)
+	} else {
+		metricJobsCurrent.With(prometheus.Labels{druidmetrics.EtcdNamespace: etcd.Namespace}).Set(0)
+	}
+
```

Comment:
Seems like a lot of code is jumbled up in the current code for compaction-controller:
The current sequence (without this PR) looks like this:
- if diff in full and delta events has crossed configured threshold, then:
  - try and fetch running compaction job
  - if it's in deletion, requeue after 10 seconds
  - if no compaction job exists, and if snapshot compaction is enabled, then deploy a new compaction job
  - if it's failed, delete and requeue the job recreation
  - if it's succeeded, delete and exit without requeueing

This logic is flawed due to the following reasons:
1. when a compaction job succeeds, the diff between full and delta events drops to near-zero, so when the requeue from the previous reconciler runs, it can never run `reconcileJob` method since the diff is no longer greater than the threshold. This means that a successful compaction job run can never be recorded in prometheus.
2. when a compaction job was previously deployed but is no longer running now, and in the meantime a regular scheduled full snapshot was taken by backup sidecar, the diff between full and delta events becomes near-zero, and thus `reconcileJob` method never runs. Due to this, a failed/successful/completed compaction job run might not be recorded in prometheus. This condition is fixed by you in the current PR, so thank you for that.
3. checking whether snapshot compaction is enabled or not should happen even before registering the compaction-controller with druid manager, and this fix would be taken care of in #671 by @seshachalam-yv 

@abdasgupta I would request you to discuss this with @seshachalam-yv so that you both can come up with a holistic solution that ensures that the logic is sound and the exported metrics are precise.

Reply:
Raised https://github.com/gardener/etcd-druid/issues/685 in favour of the discussion